### PR TITLE
Update email upsell button to link directly to email addition page

### PR DESCRIPTION
**Pull request by wongasy**
#### Proposed Changes

Currently, the **"Create a new mailbox"** button in the Embedded Inbox links out to **Upgrades > Email** `/email/:site`, where users will need to make two more clicks (select a domain, click on "Add new mailboxes") before they get to the email addition screen:

Screen Shot 2022-06-28 at 3 20 00 PM

The proposed change improves UX by linking users **directly to the email addition page**, provided that there is **a single domain** associated with the domain.

Screen Shot 2022-06-28 at 3 20 18 PM

Also took the opportunity to separate out `NewMailboxUpsell` into its own file, so things look cleaner with the additional logic for generating a different click through URL for the button. 

Ideally, I would love to have `<NewMailboxUpsell>` added via `/inbox/index.jsx` rather than adding it as part of `<MailboxSelectionList>`, since it isn't really a mailbox list item and can be placed anywhere on the page. But for now, I left upsell notice as-is, since we'd want the notice to appear at the same time as the mailbox list.

#### Testing Instructions

**Scenario A:** User has multiple domains associated with the site
1. Have multiple domains
2. Subscribe to an email service
3. Click on **"Inbox"** in the sidebar
4. Click on **"Create a new mailbox"** button

**Expected outcome:** Users land on the **Upgrades > Email** page `/email/:site`, where they need to select a domain.

---

**Scenario B:** User has a single domain associated with the site, and is subscribed to Professional Email
1. Have a single domain
2. Subscribe to Professional Email service
3. Click on **"Inbox"** in the sidebar
4. Click on **"Create a new mailbox"** button

**Expected outcome:** Users land on the email addition page for Titan `/email/:domain/titan/new/:site?source=inbox`

---

**Scenario C:** User has a single domain associated with the site, and is subscribed to Google Workspace
1. Have a single domain
2. Subscribe to Google Workspace service
3. Click on **"Inbox"** in the sidebar
4. Click on **"Create a new mailbox"** button

**Expected outcome:** Users land on the email addition page for Google `/email/:domain/google-workspace/add-users/:site?source=inbox`

### CODE REVIEW

File: client/my-sites/email/inbox/mailbox-selection-list/index.jsx
Comment (wongasy) [inline client/my-sites/email/inbox/mailbox-selection-list/index.jsx]: Getting the `domain` passed from the [mailbox object](https://github.com/Automattic/wp-calypso/pull/65058/files#diff-826f692b78a2fbee078b7b3e0372cb4c986011d0f98d577351077a2a68d64438R280) instead of getting it from `selectedSite` because the free domain gets returned as `selectedSite.domain` in **Scenario C** (free domain is set as primary address despite having a custom domain in place), and we end up with an invalid URL `/email/:site/manage/:site`.

**Comment by fredrikekelund** [comment]
There's also a scenario where a site has multiple custom domains, in which case I think we should fall back to the previous behavior (linking to `/email/:site_slug`).

File: client/my-sites/email/inbox/mailbox-selection-list/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/mailbox-selection-list/index.jsx]: I think it would be nice to take this opportunity to pass this path as an `href` prop to `Button` instead of using the `page` function to navigate programmatically.

We could still use the `onClick` handler for tracking the click event.

File: client/my-sites/email/inbox/mailbox-selection-list/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/mailbox-selection-list/index.jsx]: Seeing as we're using the `mailboxes` array to extract the domain, I think we could change the value here to `mailboxes[ 0 ]?.domain`. That way we don't have to pass the prop explicitly to this component.

**Review by fredrikekelund — CHANGES_REQUESTED** [review]
This is a nice UX improvement! 👍 

I left a couple of comments. The most important thing is probably that we should consider a scenario where a user has multiple custom domains connected to a single site.

**Comment by olaseni** [comment]
> The most important thing is probably that we should consider a scenario where a user has multiple custom domains connected to a single site.

Perhaps we could borrow the approach from #64933 and make a case exclusively for when the user has a single custom domain in that site?

File: client/my-sites/email/inbox/mailbox-selection-list/index.jsx
Comment (olaseni) [reply client/my-sites/email/inbox/mailbox-selection-list/index.jsx]: Note that this solution also only works for Professional Email, and not Google Workspace.

**Comment by wongasy** [comment]
Thank you @olaseni and @fredrikekelund for the reviews!  Great catch on the use cases for multiple domains and Google Workspace! I ended up making a fair amount of changes to address those scenarios, and also did a bit of refactoring. Since there are quite a lot of changes (compared to my initial commit), I re-wrote the PR as well :)

File: client/my-sites/email/inbox/mailbox-selection-list/index.jsx
Comment (wongasy) [reply client/my-sites/email/inbox/mailbox-selection-list/index.jsx]: I left `page` in place. I don't actually know what `page` does 😆 but I wonder if it applies some sort of delay to ensure our tracking event gets sent before navigating away from the current page? Is `page` a third party library?

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: Kudos for going the classic WordPress route here and choosing a Yoda conditional 😉

I know that the [JS coding guidelines](https://wpcalypso.wordpress.com/devdocs/docs/coding-guidelines/javascript.md#yoda-conditions) are ambivalent on the matter, but my experience is that they are in fact quite rare in Calypso and for the back-end, there was a [recent change](https://make.wordpress.org/core/2022/06/14/upcoming-disallow-assignments-in-conditions-and-remove-the-yoda-condition-requirement-for-php/) in the guidelines where they are now actively discouraged (the article provides some nice background as well).

I would encourage you not to use them, but it's no requirement.

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: We should use the `hasTitanMailWithUs` function here (located in `calypso/lib/titan`)

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: We should use the `hasGSuiteWithUs` function here (located in `calypso/lib/gsuite`)

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: Seeing as `MailboxSelectionList` is [only rendered if there is at least one domain](https://github.com/Automattic/wp-calypso/blob/af50f9d5b36fb6e0b63fa7eae56afdbc099dff26/client/my-sites/email/inbox/index.jsx#L121), I don't think we actually need to use optional chaining after `domains[ 0 ]`, right?

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: I think it would improve readability to either make this an `else` condition of the `1 === domains.length` conditional, or alternatively to make this assignment to the `upsellURL` variable the default one.

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: I know this `useCallback` call was here previously, but I actually think we could take this opportunity to drop it. There are no performance gains from using it here and it increases complexity.

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: I know this also is identical to what we had previously, but this string just sounds a bit wonky to me. Should we change it to something like `Create a new one and activate it immediately`.

@stephanethomas @AllTerrainDeveloper @olaseni WDYT?

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: I still think it would be nice to use an `href` prop here 😊

`page` is this tiny client-side routing library that we use in Calypso. The main function can either define routes (call it with one string and one callback) or navigate to routes (call it with just a string). Here's the [API reference](https://www.npmjs.com/package/page#api).

If we preserve the `onClick` handler (but remove the `page` call in that function) and add an `href` prop – I can confirm that the same events are tracked in Calypso.

File: client/my-sites/email/inbox/new-mailbox-upsell/style.scss
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/style.scss:4]: Nice adaptation of the class names 👍

**Review by fredrikekelund — CHANGES_REQUESTED** [review]
Making very good progress here! 👍

I left a number of comments, but most of them should hopefully be quite straightforward to address.

Moving `NewMailboxUpsell` into its own file also seems like a reasonable decision now that it's grown a bit.

File: client/my-sites/email/inbox/mailbox-selection-list/index.jsx
Comment (fredrikekelund) [reply client/my-sites/email/inbox/mailbox-selection-list/index.jsx]: I followed up in a new thread!

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (wongasy) [reply client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: Agreed! Especially when we're dropping `page()` and adding the `href` to the button itself.

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (wongasy) [reply client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: Thanks for keeping me updated on the guidelines :) Very helpful!

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (wongasy) [reply client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: I agree that the suggested copy sounds better!

**Comment by wongasy** [comment]
@fredrikekelund I have addressed all your feedbacks :) Thank you again for the review, all suggestions were good!

I also made another UX update so that the "Back" and "Cancel" button in the email addition screen aligns with the context of how user got there. If the `source` is `inbox`, then `goToEmail()` should link users back to the **Inbox** page as opposed to the **Email Management** page, given the **Inbox** is actually where the user came from.

File: client/my-sites/email/gsuite-add-users/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/gsuite-add-users/index.jsx]: This is a good initiative, but with the very recent completion of the "unify mailboxes" project, this component is no longer in use.

You should rebase this branch and move this implementation to the `goToEmail` function in `client/my-sites/email/add-mailboxes/index.tsx`.

File: client/my-sites/email/titan-add-mailboxes/index.jsx
Comment (fredrikekelund) [inline client/my-sites/email/titan-add-mailboxes/index.jsx]: Same here as for the `GSuiteAddUsers` component 🙂

**Comment by fredrikekelund** [comment]
@wongasy good idea with the modified logic around the cancel/back operation! I left some comments about the implementation, though.

Hopefully we should be good to merge after you've addressed those 👍

File: client/my-sites/email/inbox/new-mailbox-upsell/style.scss
Comment (olaseni) [inline client/my-sites/email/inbox/new-mailbox-upsell/style.scss:27]: Why do we need this hard width?

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (olaseni) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]:
```suggestion
	const getSectionSlug = ( domain ) => {
		if ( hasTitanMailWithUs( domain ) ) {
			return 'titan/new';
		}

		if ( hasGSuiteWithUs( domain ) ) {
			return 'google-workspace/add-users';
		}	
		
		return null
	};

	if ( domains.length === 1 && getSectionSlug( domain[ 0 ] ) ) {
		upsellURL = emailManagementEdit( selectedSiteSlug, domains[ 0 ].domain, getSectionSlug( domain [ 0 ] ), null, {
				source: INBOX_SOURCE,
			} );
	}
```

Nit: would this be more readable? Less logic duplication?

**Review by olaseni — CHANGES_REQUESTED** [review]
This needs to be rebased like @fredrikekelund mentioned as referenced files no longer exist, then we should probably test this again.

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (olaseni) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: I forgot to mention - we now always add TS/TSX files rather than JSX.

File: client/my-sites/email/inbox/new-mailbox-upsell/style.scss
Comment (wongasy) [reply client/my-sites/email/inbox/new-mailbox-upsell/style.scss:27]: Without this, the upsell CTA doesn't line up properly with the mailbox list: https://github.com/Automattic/wp-calypso/blob/8ca55b87d2292b4ece8d2b2549f7103ff5c1afa2/client/my-sites/email/inbox/mailbox-selection-list/style.scss#L123

File: client/my-sites/email/inbox/new-mailbox-upsell/index.jsx
Comment (wongasy) [reply client/my-sites/email/inbox/new-mailbox-upsell/index.jsx]: I like the idea of calculating a `slug` and removed the duplicated `emailManagementEdit()` :) I've adapted a similar approach, but without introducing a new helper function `getSectionSlug()`. I'd like to avoid having to call it twice, when we could have just stored the calculated value in a variable.

**Comment by wongasy** [comment]
For some reason, the build is failing at `Unit tests`. I'm not seeing any explicit message that explains what went wrong... I've tried rebuilding a few times without success. Wonder if this was because I renamed the `email/inbox/new-mailbox-upsell/index.jsx` to `.tsx`?

**Comment by fredrikekelund** [comment]
> For some reason, the build is failing at Unit tests. I'm not seeing any explicit message that explains what went wrong...

@wongasy I also had some trouble deciphering the TeamCity logs to begin with, but this is a type error in `client/my-sites/email/inbox/new-mailbox-upsell/index.tsx` that you can find some more details about under the Build Log tab ([ref](https://teamcity.a8c.com/viewLog.html?buildId=8164173&buildTypeId=calypso_RunAllUnitTests&tab=buildLog&_focus=480)).

File: client/my-sites/email/inbox/new-mailbox-upsell/index.tsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.tsx]: This should be `slug = 'google-workspace/add-users';`, right?

File: client/my-sites/email/inbox/new-mailbox-upsell/index.tsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.tsx]: Nit, but if we make the default value an empty string instead of `null`, Typescript will automatically assign the type `string` to the variable, which should be appropriate.

File: client/my-sites/email/inbox/new-mailbox-upsell/index.tsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.tsx]: Nit, but the notation `ResponseDomain[]` is probably more common compared to `Array< ResponseDomain >` (at least in Calypso). So long as we don't have any specific reason for using `Array<>` (e.g. we are defining multiple allowed types like `Array< string | null >`), I think it makes sense to go with `ResponseDomain[]`.

File: client/my-sites/email/inbox/new-mailbox-upsell/index.tsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.tsx]: Another nit, but why the backslash in this comment? 🤔

**Review by fredrikekelund — CHANGES_REQUESTED** [review]
Back button from the "add mailbox" page works well, `href` is in place on the button 👍 It looks good, but I found a few nitpicks and one blocker (variable name) before we can merge.

File: client/my-sites/email/inbox/new-mailbox-upsell/index.tsx
Comment (fredrikekelund) [inline client/my-sites/email/inbox/new-mailbox-upsell/index.tsx]: I missed this one previously, but since we have the `domainItem` variable here, we should probably take the opportunity to change this to `hasTitanMailWithUs( domainItem )` (same thing for `hasGSuiteWithUs`)

**Review by fredrikekelund — APPROVED** [review]

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7419598

Thank you @wongasy for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7419598).

**Pull request merged**
